src/server/pkg/storage/chunk/reader.go

```diff
@@ -104,6 +119,42 @@ func (r *Reader) readChunk(chunk *Chunk) error {
 	return nil
 }
 
+// WriteToN writes n bytes from the reader to the passed in writer. These writes are
+// data reference copies when full chunks are being written to the writer.
+func (r *Reader) WriteToN(w *Writer, n int64) error {
+	for {
+		// Read from the current data reference first.
+		if r.r.Len() > 0 {
+			nCopied, err := io.CopyN(w, r, int64(math.Min(float64(n), float64(r.r.Len()))))
```

Comment:
Is it possibly worth it to use `io.CopyBuffer` so as to avoid allocating a new buffer?

Reply:
It might help a little bit, but this code is significantly different now in the current iteration of copying. I think I will pass on this for now.

Reply:
Makes sense.